docs/badges/configuration.rst

```diff
@@ -0,0 +1,250 @@
+Configuration
+=============
+
+Badges feature configuration includes:
+
+- feature configuration
+- badges management
+- requirements setup
+- backends configuration (optional)
+
+
+Feature configuration
+---------------------
+
+Badges are under a feature flag (disabled by default):
+
+.. code-block:: sh
+
+    # credentials settings:
+    ENABLE_BADGES = false
+
+Also the feature has its configuration which may look like this:
+
+.. code-block:: sh
+
+    # credentials settings:
+    BADGES_CONFIG = {
+        "processor": "credentials.apps.badges.processing.BadgeProcessor",
+        "collectors": [
+            "credentials.apps.badges.collecting.BadgeCollector",
+            "credentials.apps.badges.backends.credly.CredlyBadgeCollector",
+        ],
+        "events": {
+            "incoming": {
+                "types": [
+                    "org.openedx.learning.course.grade.passed.v1",
+                    "org.openedx.learning.course.grade.failed.v1",
+                ]
+            },
+            "outgoing": {
+                "topics": ["badges"],
+                "types": [
+                    "org.openedx.learning.badge.awarded.v1",
+                    "org.openedx.learning.badge.revoked.v1",
+                ]
+            }
+        }
+    }
+
+Configuration describes:
+
+- `processor`_ - implementation of incoming events processing pipeline.
+- `collector`_ - implementation of badging progress collecting and side effects.
+- events - explicit configurations for *incoming* and *outgoing* events respectively:
+    - ``incoming`` section is used to auto-register handlers for a subset of expected event types;
+    - ``outgoing`` section is used to auto-register public signals in event bus producer;
+
+Incoming Events configuration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Listed incoming event types become available for requirements configuration.
+2. Listed incoming event types are taken into account during processing - `Badge Processor`_ is auto-connected to listen to those event types.
+
+.. code-block:: sh
+
+    BADGES_CONFIG["events"]["incoming"]["types"] = [
+        "org.openedx.learning.course.grade.passed.v1",
+        "org.openedx.learning.course.grade.failed.v1",
+        ...
+    ]
+
+Outgoing Events configuration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Badges configuration lists event types which must be registered for external emission.
+
+.. code-block:: sh
+
+    BADGES_CONFIG["events"]["outgoing"]["types"] = [
+        "org.openedx.learning.badge.awarded.v1",
+        "org.openedx.learning.badge.revoked.v1",
+    ]
+
+Additionally, outgoing events configuration manages event bus target topics to push events to:
+
+.. code-block:: sh
+
+    BADGES_CONFIG["events"]["outgoing"]["topics"] = ["badges"]
+
+By default these event types are dynamically auto-registered to EventBus producer in addition to already present in `EVENT_BUS_PRODUCER_CONFIG`.
```

Comment:
This is very convenient but I could see it also being surprising when trying to debug and understand where events are coming from.  Walking into credentials from other apps, you would not expect other config to be auto-registering events.  The flip-side of having to do this manually is not great either as that could lead to a different set of errors during setup.  Not a blocker but I wish there was a better way for this to be explicit without making setup more error prone.

At the very least when we add the relevant code, we should make sure to add comments near where the event bus defaults are configured to so that operators know that some events come from the BADGES_CONFIG as well.

Reply:
Yeah... I still doubt it. I don't like implicit events extension either.

Technically, there are no required events for the feature - we can simplify everything:
- no implicit auto-registration for badges-specific events in favor of manual explicit event-bus producer registration for new badge-events;
- there will be incoming event types filter only;

I'll update the draft later. Everything should become more transparent then.